SmbusI801.p

```diff
@@ -632,325 +632,196 @@ unlock:
-public NTSTATUS:ioctl_i801_write_byte_data(in[], in_size, out[], out_size) {
-    if (in_size < 3)
-        return STATUS_BUFFER_TOO_SMALL;
+    // Read the PCI vendor/device ID
+    new pci_ids;
+    status = pci_config_read_dword(pci_addr[0], pci_addr[1], pci_addr[2], 0x00, pci_ids);
+    if (!NT_SUCCESS(status))
+        return status;
 
-    new address = in[0];
-    new command = in[1];
-    new data = in[2];
+    // Read the PCI subsystem vendor/device ID
+    new pci_subsys_ids;
+    status = pci_config_read_dword(pci_addr[0], pci_addr[1], pci_addr[2], 0x2C, pci_subsys_ids);
+    if (!NT_SUCCESS(status))
+        return status;
 
-    new unused;
+    out[2] = pci_ids | (pci_subsys_ids << 32);
 
-    return i801_access_simple(address, I2C_SMBUS_WRITE, command, I2C_SMBUS_BYTE_DATA, data, unused);
+    return STATUS_SUCCESS;
 }
 
-/// SMBus word read.
-///
-/// The SMBus Read/Write Word protocol (SMBWord) transfers 2 bytes of data.
-/// This protocol also uses a command value to reference up to 256 word-sized
-/// virtual device registers.
+/// Set the SMBus clock frequency.
 ///
-/// @param in [0] = Address, [1] = Command
-/// @param in_size Must be 2
-/// @param out [0] = Data
+/// @param in [0] = Frequency in Hz or -1 for no change
+/// @param in_size Must be 1
+/// @param out [0] = Previous frequency in Hz
 /// @param out_size Must be 1
+/// @note The i801 SMBus controller use a fixed clock frequency of 100kHz.
 /// @return An NTSTATUS
-/// @warning You should acquire the "\BaseNamedObjects\Access_SMBUS.HTP.Method" mutant before calling this
-forward NTSTATUS:ioctl_i801_read_word_data(in[], in_size, out[], out_size);
-public NTSTATUS:ioctl_i801_read_word_data(in[], in_size, out[], out_size) {
-    if (in_size < 2)
-        return STATUS_BUFFER_TOO_SMALL;
+forward NTSTATUS:ioctl_clock_freq(in[], in_size, out[], out_size);
+public NTSTATUS:ioctl_clock_freq(in[], in_size, out[], out_size) {
     if (out_size < 1)
         return STATUS_BUFFER_TOO_SMALL;
-
-    new address = in[0];
-    new command = in[1];
-
-    return i801_access_simple(address, I2C_SMBUS_READ, command, I2C_SMBUS_WORD_DATA, 0, out[0]);
-}
-
-/// SMBus word write.
-///
-/// The SMBus Read/Write Word protocol (SMBWord) transfers 2 bytes of data.
-/// This protocol also uses a command value to reference up to 256 word-sized
-/// virtual device registers.
-///
-/// @param in [0] = Address, [1] = Command, [2] = Data
-/// @param in_size Must be 3
-/// @param out Unused
-/// @param out_size Unused
-/// @return An NTSTATUS
-/// @warning You should acquire the "\BaseNamedObjects\Access_SMBUS.HTP.Method" mutant before calling this
-forward NTSTATUS:ioctl_i801_write_word_data(in[], in_size, out[], out_size);
-public NTSTATUS:ioctl_i801_write_word_data(in[], in_size, out[], out_size) {
-    if (in_size < 3)
+    if (in_size < 1)
         return STATUS_BUFFER_TOO_SMALL;
 
-    new address = in[0];
-    new command = in[1];
-    new data = in[2];
+    out[0] = 100000;
```

Comment:
Actually, the current code here suffers from the issue I just mentioned in my previous comment: `out` is backed by the same buffer a s `in`. The only reason this works by chance is that the out buffer starts 32 bytes earlier because the input ioctl has the symbol name. If you wrote `out[4]` here then the next read would just read out `100000`.